flowinfra: fix shutting down vectorized flows with wrapped processors

yuzefovich opened the pull request:
flowinfra: fix shutting down vectorized flows with wrapped processors

Our flow shutdown mechanism relies on all components to listen to the
same context which is canceled by the "root" component (either
materializer or outbox) when the query needs to exit. Columnarizers
currently capture the context that is passed into their constructor
and will pass that context further into their downstream operators.
However, I believe it is possible that the cancellation of the "flow"
context will not trigger the cancellation of the columnarizer's context,
so some components are left hanging.

The problem was that we derived a new context in `startInternal` call,
but it was not returned outside of `FlowBase.Setup` method, so
ServerImpl.setupFlow used to return an old context. Now this is fixed by
doing the context derivation as the first step of `Setup` and returning
it.

Fixes: #43269.

Release note (bug fix): Previously, a query shutdown mechanism could fail
to fully cleanup the infrastructure when the query was executed via the
vectorized engine and the query plan contained wrapped row-by-row
processors (in 19.2 release those are Lookup join and Index join).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43579)
<!-- Reviewable:end -->

yuzefovich committed:
flowinfra: fix shutting down vectorized flows with wrapped processors

Our flow shutdown mechanism relies on all components to listen to the
same context which is canceled by the "root" component (either
materializer or outbox) when the query needs to exit. Columnarizers
currently capture the context that is passed into their constructor
and will pass that context further into their downstream operators.
However, I believe it is possible that the cancellation of the "flow"
context will not trigger the cancellation of the columnarizer's context,
so some components are left hanging.

The problem was that we derived a new context in `startInternal` call,
but it was not returned outside of `FlowBase.Setup` method, so
ServerImpl.setupFlow used to return an old context. Now this is fixed by
doing the context derivation as the first step of `Setup` and returning
it.

Release note (bug fix): Previously, a query shutdown mechanism could fail
to fully cleanup the infrastructure when the query was executed via the
vectorized engine and the query plan contained wrapped row-by-row
processors (in 19.2 release those are Lookup join and Index join).

asubiotto reviewed (commented):
Did you figure out how the context passed into the constructor differed from the expected context? Would be nice to figure that out before getting this in just to understand this problem more. Also, could you add a wrapped processor to the `VectorizedFlowShutdownTest` to test this scenario? I guess it's specific to how a flow is set up and the test does that itself, but might be good to have that condition anyway.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-:-LyUll2C9nPEl5jEppVq:bplxkio)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Hm, I spent some time trying to understand where the contexts deviate to no success. I came to this solution from "philosophical" grounds. I feel like there might be a place in the code where we don't use the "descendant" of the cancellation context, but I couldn't find it. One place where this could happen maybe is in the flow scheduler, I'm not very familiar with that code though.

I also spent some time thinking about how to create a unit test for this and also didn't succeed. The complication here is that if we explicitly use different contexts for wrapped processors, then we're not doing what the "production" code now will do; but if we use the same contexts, then we're not testing much. It seems like too many things to mock out. I added the logic test from the issue, and it seems to me to be the best option.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-:-LyVKKBzBuxsXZktNY_h:b-ca2kkl)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/flowinfra/flow.go, line 339 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-LyVMRWs8QTeoo_ecgcn:-LyVMRWs8QTeoo_ecgco:b-e24myt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f946b8bb62629002c5e958373b81ca9e1920dd6/pkg/sql/flowinfra/flow.go#L339)):*
> ```Go
> // Start is part of the Flow interface.
> func (f *FlowBase) Start(ctx context.Context, doneFn func()) error {
> 	if _, err := f.startInternal(ctx, f.processors, doneFn); err != nil {
> ```

I thought that maybe this spot was the culprit, so I tried returning new context from `Start` method - unfortunately, that didn't work.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Isn't the problem the fact that we only derive a cancellable context in `startInternal` but are very likely to use a parent context in `Setup` since none is returned from `Setup`? How about rather than using `SetupCtxCancel`, we derive a cancellable context in `Setup`, store that in the flow, and then do the same in `startInternal`, then set `ctxCancel` to cancel both?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-:-LyVcO4aC5A_69Cl-cXG:bhhz9fh)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
So you're suggesting having two `ctxCancel`s in `FlowBase`? That would also mean two `ctxDone`s.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-:-LyVipIe4w77Hg2uXq4D:b-ea0k9w)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Oh, I think I see the problem.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-:-LyVlGGu2gVi5XprRj_f:b2r9zw9)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

asubiotto commented:
I'm suggesting having a `setupCtxCancel` and `setupCtxDone` that is set in `Setup` and then in `startInternal`:
```
ctx, cancel = WithCancel(startInternalCtx)
flow.ctxCancel = func () {
    flow.setupCtxCancel()
    cancel()
    <-flow.setupCtxDone
}
flow.ctxDone = ctx.Done()
```
Or something like that

yuzefovich reviewed (commented):
Yep, that was it. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-:-LyVnbVUATMB7bBEnhMQ:b2ipi6d)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsql/server.go, line 352 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-LyVnUOUD-ysHCKuzP93:-LyVnUOUD-ysHCKuzP94:b-6wwm3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/777b798d8e32c7baf0a0f6a7184e216c6cc94296/pkg/sql/distsql/server.go#L352)):*
> ```Go
> 	}
> 	var err error
> 	if ctx, err = f.Setup(ctx, &req.Flow, opt); err != nil {
> ```

This was the actual problem - we were not returning the correct context from `ServerImpl.setupFlow`.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r2, 4 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43579#-:-LyVoPShA7SYastsmmcR:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

yuzefovich commented:
Flake.

bors r+
